source/server/listener_impl.cc

```diff
@@ -536,5 +636,68 @@ void ListenerImpl::setSocketFactory(const Network::ListenSocketFactorySharedPtr&
   socket_factory_ = socket_factory;
 }
 
+bool ListenerImpl::supportUpdateFilterChain(const envoy::config::listener::v3::Listener& config,
+                                            bool worker_started) {
+  // It's adding very little value to support filter chain only update at start phase.
+  if (!worker_started) {
+    return false;
+  }
+
+  // Currently we only support TCP filter chain update.
+  if (Network::Utility::protobufAddressSocketType(config.address()) !=
+          Network::Address::SocketType::Stream ||
+      Network::Utility::protobufAddressSocketType(config.address()) !=
+          Network::Address::SocketType::Stream) {
+    return false;
+  }
+
+  // Guard the filter_chains[0] access.
+  if (config.filter_chains_size() == 0) {
+    return false;
+  }
+
+  // See buildProxyProtocolListenerFilter().
+  if (PROTOBUF_GET_WRAPPED_OR_DEFAULT(config_.filter_chains()[0], use_proxy_proto, false) ^
+      PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.filter_chains()[0], use_proxy_proto, false)) {
+    return false;
+  }
+
+  // See buildTlsInspectorListenerFilter().
+  if (needTlsInspector(config_) ^ needTlsInspector(config)) {
+    return false;
+  }
+  return ListenerMessageUtil::filterChainOnlyChange(config_, config);
+}
+
+ListenerImplPtr
+ListenerImpl::newListenerWithFilterChain(const envoy::config::listener::v3::Listener& config,
+                                         bool workers_started, uint64_t hash) {
+  return std::make_unique<ListenerImpl>(
+      *this, config, version_info_, parent_, name_, added_via_api_,
+      /* new new workers started state */ workers_started,
+      /* use new hash */ hash, parent_.server_.options().concurrency());
+}
+
+void ListenerImpl::diffFilterChain(const ListenerImpl& listener,
+                                   std::function<void(Network::DrainableFilterChain&)> callback) {
+  for (const auto& filter_chain : filter_chain_manager_.allFilterChains()) {
+    if (listener.filter_chain_manager_.allFilterChains().find(filter_chain.first) ==
+        listener.filter_chain_manager_.allFilterChains().end()) {
+      // exists in this listener but not in other
+      callback(*filter_chain.second);
+    }
+  }
+}
+
+bool ListenerMessageUtil::filterChainOnlyChange(const envoy::config::listener::v3::Listener& lhs,
+                                                const envoy::config::listener::v3::Listener& rhs) {
+  Protobuf::util::MessageDifferencer differencer;
+  differencer.set_message_field_comparison(Protobuf::util::MessageDifferencer::EQUIVALENT);
+  differencer.set_repeated_field_comparison(Protobuf::util::MessageDifferencer::AS_SET);
+  differencer.IgnoreField(
+      envoy::config::listener::v3::Listener::GetDescriptor()->FindFieldByName("filter_chains"));
+  return differencer.Compare(lhs, rhs);
+}
```

Comment:
This is the major function determine whether we can update the listener without draining all the existing connections